### PR TITLE
perf(agents): parallel charter discovery with bounded concurrency

### PR DESCRIPTION
**Pull request by spboyer**
## Summary

Three serial `for...await` loops in agent/charter discovery were linear in the number of agents — costly on GitHub-sourced teams (each iteration is an HTTP round-trip) and wasteful on local storage even though FS reads are fan-out safe up to ulimit.

For an 8-agent team this drops `GitHubAgentSource.listAgents()` latency from ~1.6-4 s of serial HTTP round-trips to ~0.4-1 s parallel.

### New utility

`packages/squad-sdk/src/utils/map-with-limit.ts`:
- `mapWithLimit(items, limit, fn)` — bounded parallel map, **preserves input order**, fast-fails on first rejection.
- `mapWithLimitSettled(items, limit, fn)` — same shape but returns per-item `PromiseSettledResult` so one bad input does not abort the batch.

Both are implemented with a worker-pool pattern and an indexed result array (not push-on-completion) so order is preserved regardless of completion sequence.

### Call sites

**`packages/squad-sdk/src/config/agent-source.ts`**:

- `LocalAgentSource.listAgents()`:
  - SquadState branch uses `mapWithLimitSettled(items, 8, ...)` — the prior implementation already `continue`d on per-agent errors, so settled semantics preserve behavior.
  - Storage fallback uses `mapWithLimit(items, 8, ...)` — fast-fail because the original code propagated `isDirectory()` / `read()` errors to the caller. Storage backend outages now still surface instead of producing a silently-partial list.
  - Storage fallback also parallelises the `isDirectory` probe so concurrency budget is spent on real work.

- `GitHubAgentSource.listAgents()`:
  - Uses `mapWithLimit(dirs, 5, ...)` — bounded at 5 to stay well clear of GitHub secondary rate limits while still removing serial HTTP latency.
  - **Critical fix from initial review**: uses `mapWithLimit` not `mapWithLimitSettled`, so a 403 / rate-limit / auth / network failure still rejects the call instead of being silently swallowed. The original `for...await` propagated these errors; the parallel version must too.

**`packages/squad-sdk/src/agents/index.ts`**:

- `CharterCompiler.compileAll()` (both `SquadState` and storage fallback branches) uses `mapWithLimitSettled(candidates, 8, ...)`. The prior implementation `try/catch`ed each charter compile and `continue`d on failure, so per-item settled semantics are correct here.

### Concurrency limits

| Site | Limit | Rationale |
|---|---|---|
| `LocalAgentSource` (FS) | 8 | Headroom under typical macOS (~256) / Linux (~1024) ulimits |
| `GitHubAgentSource` (HTTP) | 5 | Comfortably below GitHub secondary rate-limit thresholds |
| `CharterCompiler.compileAll` | 8 | Same FS rationale |

## Testing

- `test/map-with-limit.test.ts` (new, 12 cases):
  - Order preservation under reverse-order completion
  - Real concurrency cap proved via gated `deferred()` (3 in flight is observable before release)
  - `mapWithLimit` fast-fails; `mapWithLimitSettled` never throws
  - Empty input, `limit > items.length`, `limit < 1` / NaN edge cases
- `test/agent-source-github.test.ts` — 19 existing tests still pass unchanged + 2 new regression tests:
  - "propagates a thrown fetcher error instead of silently dropping the agent" (asserts 403 from a mocked fetcher rejects the call — would have caught the bug fixed during review)
  - "still skips agents whose charter.md returns null" (preserves the existing missing-charter behavior)
- `test/charter-compiler.test.ts` — 38 existing tests pass unchanged.

## Notes

Part of a five-PR perf series. Companion PRs:
- chore: dependency hygiene
- feat(bench): npm run bench + bench:cold-start runners
- resolution cache + multi-squad config dedupe
- non-blocking scheduler script execution

### CODE REVIEW

File: packages/squad-sdk/src/utils/map-with-limit.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * @returns Array of results in the same order as `items`.
+ */
+export async function mapWithLimit<T, R>(
+  items: readonly T[],
+  limit: number,
+  fn: (item: T, index: number) => Promise<R>,
+): Promise<R[]> {
+  if (limit < 1 || !Number.isFinite(limit)) {
+    throw new Error(`mapWithLimit: limit must be a positive integer, got ${limit}`);
+  }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply packages/squad-sdk/src/utils/map-with-limit.ts:34]: Fixed in b7310ca6 — strengthened both `mapWithLimit` and `mapWithLimitSettled` to use `!Number.isInteger(limit)` instead of `!Number.isFinite(limit)`, so non-integer values like `1.5` and `Infinity` now reject with the documented "positive integer" error. Regression tests added in `test/map-with-limit.test.ts`.

File: scripts/pr-readiness.mjs [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+  const rows = displayed.map((f) => {
+    const added = f.additions || 0;
+    const deleted = f.deletions || 0;
+    return `| \`${sanitizeFilename(f.filename)}\` | +${added} −${deleted} |`;
+  });
+
+  if (truncated) {
+    const remaining = files.length - MAX_FILE_LIST;
+    rows.push(`| ... | **+${remaining} more files** | |`);
+  }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply scripts/pr-readiness.mjs]: Acknowledged — this file is out-of-scope for this perf PR (it ended up in the diff because the branch was cut from a stale base). Tracking the fix in the duplicate threads on PR #1110 / #1111 / #1114; the table-row column-count fix (`| ... | **+N more files** |`) will land via a focused repo-health PR. Will rebase this branch onto current `dev` once that lands so the file drops out of the diff here.

File: .github/workflows/squad-pr-nudge.yml [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+              // 5. Check if branch is behind base
+              const comparison = await github.rest.repos.compareCommits({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                base: pr.base.ref,
+                head: pr.head.sha
+              });
+              if (comparison.data.ahead_by > 10) {
+                actions.push(`⬇️ **${comparison.data.ahead_by} commits behind ${pr.base.ref}.** Rebase to pick up latest changes.`);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply .github/workflows/squad-pr-nudge.yml]: Acknowledged — same out-of-scope situation as the pr-readiness.mjs thread: this file is in the diff only because the branch was cut from a stale base. The `ahead_by` → `behind_by` swap (and matching message text) will land via a focused repo-health PR; this branch will be rebased afterward so the file drops out.

File: .github/workflows/squad-repo-health.yml [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    timeout-minutes: 3
+    if: github.actor != 'dependabot[bot]'
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          sparse-checkout: |
+            scripts/check-bootstrap-deps.mjs
+          sparse-checkout-cone-mode: false
+      - name: Fetch PR head (data only — not executed)
+        run: git fetch origin ${{ github.event.pull_request.head.sha }}
```
[the remark this answers is not in the document]
Comment (spboyer) [reply .github/workflows/squad-repo-health.yml]: Acknowledged — same out-of-scope situation. This workflow file is in the diff only because the branch was cut from a stale base. The fork-PR `git fetch` failure mode you describe is real; tracking the fix via a focused repo-health PR. This branch will be rebased onto current `dev` once that lands.

File: packages/squad-sdk/src/config/agent-source.ts [hunk 5]
```diff
@@ -34,6 +35,19 @@ export interface AgentDefinition extends AgentManifest {
...
+ * - LOCAL_LIST_CONCURRENCY: filesystem reads are cheap individually but the
+ *   default ulimit on file descriptors is finite. 8 leaves headroom for the
+ *   rest of the SDK while still saturating typical 5-10 agent teams.
+ * - GITHUB_LIST_CONCURRENCY: GitHub REST API has secondary rate limits on
+ *   burst concurrency; 5 stays well clear of those thresholds for typical
+ *   team sizes while removing the serial-await bottleneck.
+ */
+const LOCAL_LIST_CONCURRENCY = 8;
+const GITHUB_LIST_CONCURRENCY = 5;
+
```
[the remark this answers is not in the document]
Comment (spboyer) [reply packages/squad-sdk/src/config/agent-source.ts:50]: You are right — this PR was branched from a stale base (`main` instead of current `dev`) which pulled in 36 unrelated commits. The intended diff is just the parallel charter-discovery change in `agent-source.ts`. Plan: (1) land a focused repo-health PR to `dev` with the genuine bugs flagged across this batch (pr-readiness column count, `ahead_by`/`behind_by`, fork-PR fetch handling, `template-sync.test` working-tree mutation, `toJson` string), (2) rebase this branch onto current `dev` so the diff is just the perf change, (3) re-request review. The single in-scope code fix (`mapWithLimit` integer validation) is in b7310ca6.

**Comment by spboyer** [comment]
### PR Checkup — feedback addressed

**Code fix applied (in-scope):**
- `b7310ca6`: `mapWithLimit` / `mapWithLimitSettled` now reject non-integer limits with the documented "positive integer" error. Regression tests added.
- Changeset: `.changeset/map-with-limit-integer-validation.md` (patch to `@bradygaster/squad-sdk`)

**Out-of-scope feedback (acknowledged + resolved):**
- 3 review threads on `scripts/pr-readiness.mjs`, `.github/workflows/squad-pr-nudge.yml`, and `.github/workflows/squad-repo-health.yml` flag real bugs but those files are only in this diff because the branch was cut from a stale base. They will be addressed in a focused repo-health PR to `dev`.

**Open thread (left intentionally):**
- The PR-scope concern thread is left open until the branch is rebased onto current `dev` (will reduce diff to just the perf change in `agent-source.ts`).

@copilot-pull-request-reviewer — re-review requested once the rebase + repo-health PR land. Thanks for the careful review!

File: packages/squad-sdk/src/config/agent-source.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (spboyer) [reply packages/squad-sdk/src/config/agent-source.ts:50]: After rebasing onto current `dev`, PR #1113 is scoped to just **parallel charter discovery with bounded concurrency** (6 files changed). The cross-cutting workflow/script/CLI/template/version-bump changes the bot flagged here were artifacts of the branch having been originally cut from a stale `main`; they are no longer present in this PR. Resolving.

**Comment by spboyer** [comment]
✅ All previously-flagged review feedback has been addressed:

- **Rebased onto current `dev`** so the diff is now scoped to just **parallel charter discovery with bounded concurrency** — earlier scope-drift comments no longer apply.
- **Resolved every unresolved review thread** with an explanatory reply.
- **CI is green** (test, Policy Gates, sdk-exports-validation, samples-build).

@copilot-pull-request-reviewer — ready for another look when you get a chance. Thanks!
